[CodeGen][ObjC] Initial WebAssembly Support for GNUstep

hmelder opened the pull request:
This pull request adds initial support for compiling Objective-C to WebAssembly. I tested my changes with [libobjc2](https://github.com/gnustep/libobjc2) and the [swift-corelibs-blocksruntime](https://github.com/swiftlang/swift-corelibs-blocksruntime/pull/7).

There are two outstanding issues, which I cannot fix as deeper knowledge of the subsystems is required:
1. Symbols marked as explicitly hidden in code generation are exported
2. Clang crashes in SelectionDAG when compiling an Objective-C try/catch block with `-fwasm-exceptions`

### First Issue

Emscripten is processing the generated `.wasm` file in `emscripten.py` and checks if all exported symbols are valid javascript identifiers ([tools/js_manipulation.py#L104](https://github.com/emscripten-core/emscripten/blob/e43e97f16fd0ed82bb0a1f25d6e1b28b2c8c08c3/tools/js_manipulation.py#L104)). However, hidden symbols such as `.objc_init` are intentionally an invalid C identifier.

The core of the problem is that symbols with the `WASM_SYMBOL_NO_STRIP` attribute are exported when targeting Emscripten (https://reviews.llvm.org/D62542). This attribute is added to the symbol during relocation in  `WasmObjectWriter::recordRelocation`. So we are accidentally exporting a lot of hidden symbols and not only ones generated by ObjC CG...

I'm currently hacking around this by not exporting no-strip symbols. This is the default behaviour for Wasm.

### Second Issue

Here is a minimal example that triggers the crash. 

```objc
#include<stdio.h>

int main(void) {
	int ret = 0;
	@try {
	}
	@catch (id a)
	{
		ret = 1;
                 puts("abc");
	}

	return ret;
}
```
The following assertion is triggered:

```
clang: /home/vm/llvm-project/llvm/lib/Target/WebAssembly/WebAssemblyExceptionInfo.cpp:124: void llvm::WebAssemblyExceptionInfo::recalculate(MachineFunction &, MachineDominatorTree &, const MachineDominanceFrontier &): Assertion `EHInfo' failed.
```
Here is the crash report [main-c3884.zip](https://github.com/user-attachments/files/23677308/main-c3884.zip).

You can use `emcc` with a modified LLVM build by exporting `EM_LLVM_ROOT` before sourcing `emsdk/emsdk_env.sh`:
```sh
emcc -fobjc-runtime=gnustep-2.2 -fwasm-exceptions -c main.m
```

or just invoke clang directly:

```sh
/home/vm/llvm-build-wasm/bin/clang -target wasm32-unknown-emscripten -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -wasm-use-legacy-eh=false -mllvm -disable-lsr --sysroot=/home/vm/emsdk/upstream/emscripten/cache/sysroot -DEMSCRIPTEN -fobjc-runtime=gnustep-2.2 -fwasm-exceptions -c main.m
```

## Building libobjc2 and the BlocksRuntime

### Building the BlocksRuntime
```sh
cmake -DCMAKE_TOOLCHAIN_FILE=$EMSDK/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake   -DCMAKE_INSTALL_PREFIX=/home/vm/demo-install -DCMAKE_BUILD_TYPE=Debug -B build -G Ninja
```

### Building libobjc2
```sh
cmake -DCMAKE_TOOLCHAIN_FILE=$EMSDK/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake   -DCMAKE_INSTALL_PREFIX=/home/vm/demo-install -DBlocksRuntime_LIBRARIES=/home/vm/demo-install/lib/libBlocksRuntime.a -DBlocksRuntime_INCLUDE_DIR=/home/vm/demo-install/include/BlocksRuntime -DEMBEDDED_BLOCKS_RUNTIME=OFF -DTESTS=OFF  -B build  -DCMAKE_BUILD_TYPE=Debug  -G Ninja
```

hmelder committed:
[clang] Whitelist wasm when targeting GNUstep 2.x

hmelder committed:
[CodeGen][ObjC] Mangle public symbols for wasm

Emscripten requires that exported symbols. See
https://github.com/emscripten-core/emscripten/pull/23563.

hmelder committed:
[CodeGen][ObjC] Fix class_registerAlias_np signature

hmelder commented:
Assuming that the new WASM exception implementation implements the mandatory functions and data structure of the Itanium EH ABI correctly, not much needs to be done to get EH working with libobjc2. I just need to find the root course of the crash...

davidchisnall commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -179,8 +179,15 @@ class CGObjCGNU : public CGObjCRuntime {
+      return (StringRef("$_") + Name).str();
+    } else {
+      return (StringRef("._") + Name).str();
+    }
```
```suggestion
    StringRef prefix = "._"

    // Exported symbols in Emscripten must be a valid Javascript identifier.
    auto triple = CGM.getTriple();
    if (triple.isOSBinFormatCOFF() || triple.isOSBinFormatWasm()) {
      prefix = "$_";
    return (prefix + Name).str();
```

davidchisnall reviewed (commented):
The Objective-C bits look fine to me, the MC bit possibly should be a separate PR.

davidchisnall reviewed (commented):
Code LGTM once clang-format is happy.

We probably should have a test in tests/CodeGenObjC checking that the mangling is correct for WAsm.

hmelder commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -179,8 +179,14 @@ class CGObjCGNU : public CGObjCRuntime {
+    StringRef prefix = "._"
+
+        // Exported symbols in Emscripten must be a valid Javascript identifier.
+        auto triple = CGM.getTriple();
```
The problem is the missing semicolon in your change :)

```c
StringRef prefix = "._"
```

hmelder committed:
[CodeGen][ObjC] Add WASM symbol mangling test

dschuff commented on clang/lib/CodeGen/CGObjCGNU.cpp:
```diff
@@ -179,8 +179,16 @@ class CGObjCGNU : public CGObjCRuntime {
+
+    // Exported symbols in Emscripten must be a valid Javascript identifier.
+    auto triple = CGM.getTriple();
+    if (triple.isOSBinFormatCOFF() || triple.isOSBinFormatWasm()) {
```
The restriction on valid JS identifiers is specific to Emscripten rather than wasm as a whole. But if you want to have a common ABI across Emscripten and WASI, then this would be OK with me too.

dschuff replied:
Yeah, better than link failure I think. And given that the map style accessors aren't going away, maybe it's sufficient to just leave it at that. where if you have invalid identifiers, you just need to use that method (as opposed to trying to mangle them and change the symbol name altogether?)

dschuff replied:
@hmelder / @HendrikHuebner Would Sam's Emscripten change be useful for the objc use case?

hmelder replied:
> The restriction on valid JS identifiers is specific to Emscripten rather than wasm as a whole, so you might want to check for isOSEmscripten here rather than the bin format. But if you want to have a common ABI across Emscripten and WASI, then this would be OK with me too.

I think it does not _really_ matter whether we use `$` or `.` and I would prefer to not add more complexity into the conditional.

> We should maybe fix emscripten to deal with these symbols instead of patching here?
> [...]
> So this change would just mean that symbol are that not valid JS symbol names would not be accessible via the first method... which is an odd difference but maybe better than "link failure" ?

We mangle the symbols so that they are not directly usable by the user in a C program, or at least this was the original intend with `.` on ELF. AFAIK there is an extension that allows `$` to be used in a C identifier which is why Emscripten supports it in the first place.

You _can_ change this in Emscripten but it does not really matter for our use case.

hmelder commented:
Take `.objc_ctor` for example

https://github.com/llvm/llvm-project/blob/917e458b964418b17d969e3fa914846df4578004/clang/lib/CodeGen/CGObjCGNU.cpp#L1578-L1580

which is placed into comdat and explicitly marked as hidden here

https://github.com/llvm/llvm-project/blob/917e458b964418b17d969e3fa914846df4578004/clang/lib/CodeGen/CGObjCGNU.cpp#L1598-L1599

When compiling a basic Objective-C application with emcc (`EM_LLVM_ROOT` pointing to my patched LLVM), I get the following error message:

`emcc: error: invalid export name: "_.objc_ctor"`

Which originates from `upstream/emscripten/tools/emscripten.py`:
```py
  for e in settings.EXPORTED_FUNCTIONS:
    if not js_manipulation.isidentifier(e):
      exit_with_error(f'invalid export name: "{e}"')
```

The hidden symbol is exported because the no-strip attributed was added implicitly, causing it to be exported in 

https://github.com/llvm/llvm-project/blob/2aa345054fa2975bd9d2e1ca03195e40c95058d7/llvm/lib/MC/WasmObjectWriter.cpp#L1795-L1800

hmelder commented:
> It looks like maybe OP is running into the issue of Emscripten's hack from the ancient times where EMSCRIPTEN_KEEPALIVE (which compiles to attribute("used") and then WASM_SYM_NO_STRIP)

Exactly this is the problem.

HendrikHuebner commented:
> There is also some work to be done to support other personality functions in WebAssembly. Basic code generation with the wrong personality function works with [this](https://github.com/llvm/llvm-project/pull/171038) patch.
> 
>     1. Clang emits `_Unwind_CallPersonality` (defined in [libunwind/src/Unwind-wasm.c](https://github.com/llvm/llvm-project/blob/80ec43d455a5e47ba005112cd2b2c447bb40c42c/libunwind/src/Unwind-wasm.c#L58)) which always calls `__gxx_personality_wasm0`.
> 
>     2. WebAssembly catchpads are emitted based on the return value of `isWasmPersonality`.
> 
>     3. WasmEHPrepare hard-codes and reports a fatal error if the personality function is not `__gxx_wasm_personality_v0`
>        https://github.com/llvm/llvm-project/blob/80ec43d455a5e47ba005112cd2b2c447bb40c42c/llvm/lib/CodeGen/WasmEHPrepare.cpp#L238-L244
> 
> 
> The personality `__gxx_personality_wasm0` is implemented in [libcxxabi/src/cxa_personality.cpp](https://github.com/llvm/llvm-project/blob/80ec43d455a5e47ba005112cd2b2c447bb40c42c/libcxxabi/src/cxa_personality.cpp#L1019).
> 
> I would like to avoid defining two new personality functions for ObjC and ObjCXX on WASM. Can't we just differentiate and emit code based on the target triple?

 I don't understand why we need `_Unwind_CallPersonality` in the first place, since all it does at the moment is delegating to the (hardcoded) personality function and resetting the selector (Which could also be done inside the personality function). There needs to be more flexibility in the ABI to have different personality functions. @dschuff Do you think an ABI break involving `_Unwind_CallPersonality` would still be tolerable at this stage, such as changing the signature or just avoiding it all together and directly emitting a call to the personality function?

dschuff commented:
> I would like to avoid defining two new personality functions for ObjC and ObjCXX on WASM. Can't we just differentiate and emit code based on the target triple?

If objc follows exactly the same ABI and runtime as C++ then maybe it can just use the same personality function. I think we do want to support different personality functions, but maybe the test is which parts of the runtime code they want to share. IIRC we also have a separate personality function for using wasm EH for setjmp and longjmp (or maybe that's just a different tag?). I'd want to get @aheejin's opinion about the best design for this. But we can certainly make this work one way or another.

aheejin commented:
@dschuff
> IIRC we also have a separate personality function for using wasm EH for setjmp and longjmp (or maybe that's just a different tag?).

We don't use a personality function to handle SjLj. I think what you're talking about might be this? https://github.com/llvm/llvm-project/blob/f440b5c12df500f86f9407e625d42f92913e8367/libcxxabi/src/cxa_personality.cpp#L1029-L1030
This is used in SjLj exception handling, which I believe is a kind of EH used in x86.

aheejin commented:
@rjmccall 
> I don't know why the GNUStep runtime needs different ObjC and ObjC++ personality functions. The ObjC++ personality has to support the complete superset of exception clauses (a single call site can have handlers for both C++ and ObjC exception types). It should therefore still be usable in either pure ObjC or pure C++.
> 
> In theory, the ObjC++ personality needs to be able to distinguish ObjC and C++ exceptions, so using it instead of a single-language personality could require a less compact LSDA format. In practice, they use the exact same format, and AFAIK the exception types are always distinguished in a more subtle way. I don't know the exact details of what GNUStep does, but the Apple ObjC runtime just provides a v-table for ObjC exception RTTI objects that implements the private exception-matching virtual methods on type_info differently. (I'm not even sure LLVM _supports_ emitting a different LSDA format for different personalities.)

If ObjC can share libc++abi (with little `#ifdef`s here and there I suppose) it will be simpler and we don't need to change much (if at all) in WasmEHPrepare.

hmelder commented:
> It's more of our convention than ABI, so I don't think it's doable if we really need to, but for our libraries I think the current code works fine. You can either
> 
> 1. Modify WasmEHPrepare to generate calls to your personality function in case of ObjC, or
> 2. Create `_Unwind_CallPersonality` wrapper in your library too
> 
> If 2 is not too much burden, I think it would help make WasmEHPrepare simpler.

See https://github.com/llvm/llvm-project/pull/175202 where I completely replaced `_Unwind_CallPersonality` with a direct call to the personality function.

hmelder committed:
[CodeGen][ObjC] Use C++-based EH for WASM targets

The Wasm EH implementation in Clang pretty much hard-codes
__gxx_wasm_personality_v0 by calling the veneer function
_Unwind_CallPersonality instead of calling the personality function
directly.

While it is possible to remove _Unwind_CallPersonality and instead
generate its body in CG, this will add a couple of instructions in each
catch block. Doable, but we can also do the following:

Since we already have C++-based EH for MinGW in CGObjCGNU, reusing it
for Wasm saves us from implementing our own personality function
and objc_begin_catch/objc_end_catch functions.
